pkg/vault/test/foundry/E2eSwap.t.sol

```diff
@@ -0,0 +1,696 @@
+        vm.prank(admin);
+        feeController.setGlobalProtocolSwapFeePercentage(FIFTY_PERCENT);
+
+        vm.prank(poolCreator);
+        // Set pool creator fee to 100%, so protocol + creator fees equals the total charged fees.
+        feeController.setPoolCreatorSwapFeePercentage(pool, FixedPoint.ONE);
+    }
+
+    /**
+     * @notice Override pool created by BaseVaultTest.
+     * @dev For this test to be generic and support tokens with different decimals, tokenA and tokenB must be set by
+     * `_setUpVariables`. If this function runs before `BaseVaultTest.setUp()`, in the `setUp()` function, tokens
+     * defined by BaseTest (like dai and usdc) cannot be used. If it runs after, we don't know which tokens are used to
+     * use createPool and initPool. So, the solution is to create a parallel function to create and init a custom pool
+     * after BaseVaultTest setUp finishes.
+     */
+    function createAndInitCustomPool() internal virtual {
+        pool = _createPool([address(tokenA), address(tokenB)].toMemoryArray(), "custom-pool");
+
+        uint256[] memory initAmounts = new uint256[](2);
+        initAmounts[tokenAIdx] = poolInitAmountTokenA;
+        initAmounts[tokenBIdx] = poolInitAmountTokenB;
+
+        vm.startPrank(lp);
+        _initPool(pool, initAmounts, 0);
+        vm.stopPrank();
+    }
+
+    /**
+     * @notice Set up tokens.
+     * @dev When extending the test, override this function and set the same variables.
+     */
+    function setUpTokens() internal virtual {
+        tokenA = dai;
+        tokenB = usdc;
+    }
+
+    /**
+     * @notice Set up test variables (sender, poolCreator, pool swap fee, swap sizes).
+     * @dev When extending the test, override this function and set the same variables.
+     */
+    function setUpVariables() internal virtual {
+        sender = lp;
+        poolCreator = lp;
+
+        // If there are swap fees, the amountCalculated may be lower than MIN_TRADE_AMOUNT. So, multiplying
+        // MIN_TRADE_AMOUNT by 10 creates a margin.
+        minSwapAmountTokenA = 10 * MIN_TRADE_AMOUNT;
+        maxSwapAmountTokenA = poolInitAmountTokenA;
+
+        minSwapAmountTokenB = 10 * MIN_TRADE_AMOUNT;
+        maxSwapAmountTokenB = poolInitAmountTokenB;
+
+        // 0.0001% min swap fee.
+        minPoolSwapFeePercentage = 1e12;
+        // 10% max swap fee.
+        maxPoolSwapFeePercentage = 10e16;
+    }
+
+    /// @notice Donate tokens to vault, so liquidity tests are possible.
+    function _donateToVault() internal virtual {
+        tokenA.mint(address(vault), 100 * poolInitAmountTokenA);
+        tokenB.mint(address(vault), 100 * poolInitAmountTokenB);
+        // Override vault liquidity, to make sure the extra liquidity is registered.
+        vault.manualSetReservesOf(tokenA, 100 * poolInitAmountTokenA);
+        vault.manualSetReservesOf(tokenB, 100 * poolInitAmountTokenB);
+    }
+
+    function testDoUndoExactInNoFees__Fuzz(uint256 exactAmountIn) public {
```

Comment:
I'm wondering whether we should test this case or not.

If a pool fails because the fee is 0 but it has a guardrail it should still be ok for the most part (unless LP fees are bypassed with aggregate swap fees completely). What if a pool type fails in this case?

Reply:
Removed it!